File: Q1/tools/analyze.cpp
// analyze.cpp
// ---------------------------------------------------------------------------
// Q1 analysis tool (C++ replacement for the Python script).
//
// Reads benchmark/results.csv and writes, on stdout, the tables the report
// format asks for:
//   * runtime            median t_algo, seconds
//   * speed-up           S(P) = T1 / TP
//   * efficiency         E(P) = S(P) / P
//   * communication      t_comm as a percentage of t_algo
//   * phase breakdown    t_dist / t_compute / t_gather
//
// And writes four SVG charts into benchmark/plots/. SVG is generated directly
// (see harness.hpp) so no plotting library is needed; it embeds in a report as
// is, and converts to PNG with any standard tool if a raster is required.
//
// The value reported per configuration is the MEDIAN over trials, which is
// robust against a single slow run caused by another job sharing the node.
//
// T1 is the MPI binary at P=1, not the sequential program, so the speed-up
// measures parallel scaling of one implementation rather than comparing two
// different programs.
//
// Build: g++ -O2 -std=c++17 -o analyze analyze.cpp
// Run  : ./tools/analyze [results.csv]      (from the Q1 project root)
// ---------------------------------------------------------------------------

#include <cstdio>
#include <map>
#include <string>
#include <vector>

#include "harness.hpp"

using harness::fmt;
using harness::median;
using harness::Series;

struct Key {
    std::string size;
    int         P;
    bool operator<(const Key& o) const {
        return size != o.size ? size < o.size : P < o.P;
    }
};

struct Agg {
    double dist = 0, compute = 0, gather = 0, total = 0, comm = 0, algo = 0;
};

int main(int argc, char** argv) {
    const std::string path = (argc > 1) ? argv[1] : "benchmark/results.csv";

    auto rows = harness::readCsv(path);
    if (rows.empty()) {
        std::fprintf(stderr,
                     "no usable results at %s\nRun ./tools/run_bench first "
                     "(from the Q1 project root).\n", path.c_str());
        return 1;
    }

    // ---- gather raw samples ------------------------------------------------
    std::map<Key, std::vector<double>> vDist, vComp, vGath, vTot;
    std::vector<std::string> sizes;
    std::map<std::string, std::string> shape;
    std::vector<int> procs;

    for (auto& r : rows) {
        Key k{r["size"], std::atoi(r["P"].c_str())};
        vDist[k].push_back(std::atof(r["t_dist"].c_str()));
        vComp[k].push_back(std::atof(r["t_compute"].c_str()));
        vGath[k].push_back(std::atof(r["t_gather"].c_str()));
        vTot[k].push_back(std::atof(r["t_total"].c_str()));
        if (std::find(sizes.begin(), sizes.end(), k.size) == sizes.end()) {
            sizes.push_back(k.size);
            shape[k.size] = r["m"] + "x" + r["n"] + "x" + r["p"];
        }
        if (std::find(procs.begin(), procs.end(), k.P) == procs.end())
            procs.push_back(k.P);
    }
    std::sort(procs.begin(), procs.end());

    // ---- medians -----------------------------------------------------------
    std::map<Key, Agg> agg;
    for (auto& kv : vComp) {
        const Key& k = kv.first;
        Agg a;
        a.dist    = median(vDist[k]);
        a.compute = median(vComp[k]);
        a.gather  = median(vGath[k]);
        a.total   = median(vTot[k]);
        a.comm    = a.dist + a.gather;
        a.algo    = a.compute + a.comm;
        agg[k] = a;
    }
    auto get = [&](const std::string& s, int p, Agg& out) {
        auto it = agg.find(Key{s, p});
        if (it == agg.end()) return false;
        out = it->second;
        return true;
    };

    // ---- header ------------------------------------------------------------
    std::printf("# Q1 Benchmark Analysis - Row-Row Matrix Multiplication\n");
    std::printf("\nSource: `%s`  |  process counts:", path.c_str());
    for (int p : procs) std::printf(" %d", p);
    std::printf("  |  statistic: median over trials\n");
    std::printf("\n`t_algo` = `t_compute` + `t_comm`, where `t_comm` = "
                "`t_dist` (Bcast dims + Scatterv A + Bcast B) + `t_gather` "
                "(Gatherv C).\n");

    {
        std::vector<std::vector<std::string>> body;
        for (auto& s : sizes) body.push_back({s, shape[s]});
        harness::printTable("Matrix sizes", {"Label", "m x n x p"}, body);
    }

    std::vector<std::string> head{"Input size"};
    for (int p : procs) head.push_back("P=" + std::to_string(p));

    // ---- runtime -----------------------------------------------------------
    {
        std::vector<std::vector<std::string>> body;
        for (auto& s : sizes) {
            std::vector<std::string> row{s};
            for (int p : procs) {
                Agg a;
                row.push_back(get(s, p, a) ? fmt(a.algo, 4) : "--");
            }
            body.push_back(row);
        }
        harness::printTable("Runtime - median `t_algo` (seconds)", head, body);
    }

    // ---- speed-up ----------------------------------------------------------
    std::map<Key, double> speed;
    {
        std::vector<std::vector<std::string>> body;
        for (auto& s : sizes) {
            std::vector<std::string> row{s};
            Agg base;
            const bool haveBase = get(s, 1, base);
            for (int p : procs) {
                Agg a;
                if (haveBase && get(s, p, a) && a.algo > 0) {
                    const double v = base.algo / a.algo;
                    speed[Key{s, p}] = v;
                    row.push_back(fmt(v, 2));
                } else {
                    row.push_back("--");
                }
            }
            body.push_back(row);
        }
        harness::printTable("Speed-up  S(P) = T1 / TP", head, body);
    }

    // ---- efficiency --------------------------------------------------------
    {
        std::vector<std::vector<std::string>> body;
        for (auto& s : sizes) {
            std::vector<std::string> row{s};
            for (int p : procs) {
                auto it = speed.find(Key{s, p});
                row.push_back(it == speed.end()
                                  ? "--"
                                  : fmt(100.0 * it->second / p, 1) + "%");
            }
            body.push_back(row);
        }
        harness::printTable("Efficiency  E(P) = S(P) / P", head, body);
    }

    // ---- communication share ----------------------------------------------
    {
        std::vector<std::vector<std::string>> body;
        for (auto& s : sizes) {
            std::vector<std::string> row{s};
            for (int p : procs) {
                Agg a;
                row.push_back((get(s, p, a) && a.algo > 0)
                                  ? fmt(100.0 * a.comm / a.algo, 1) + "%"
                                  : "--");
            }
            body.push_back(row);
        }
        harness::printTable(
            "Communication share - `t_comm` as % of `t_algo`", head, body,
            "Rising with P is expected: the Bcast of B moves n*p elements to "
            "every process regardless of P, while the computation per process "
            "falls as 1/P.");
    }

    // ---- phase breakdown ---------------------------------------------------
    {
        std::vector<std::vector<std::string>> body;
        for (auto& s : sizes)
            for (int p : procs) {
                Agg a;
                if (!get(s, p, a)) continue;
                body.push_back({s, std::to_string(p), fmt(a.dist, 4),
                                fmt(a.compute, 4), fmt(a.gather, 4),
                                fmt(a.algo, 4)});
            }
        harness::printTable(
            "Phase breakdown (seconds, median)",
            {"Size", "P", "t_dist", "t_compute", "t_gather", "t_algo"}, body);
    }

    // ---- plots -------------------------------------------------------------
    std::system("mkdir -p benchmark/plots");

    auto build = [&](double (*pick)(const Agg&, double), bool useSpeed) {
        std::vector<Series> out;
        for (auto& s : sizes) {
            Series ser;
            ser.name = s + " (" + shape[s] + ")";
            for (int p : procs) {
                if (useSpeed) {
                    auto it = speed.find(Key{s, p});
                    if (it == speed.end()) continue;
                    ser.x.push_back(p);
                    ser.y.push_back(pick(Agg{}, it->second));
                } else {
                    Agg a;
                    if (!get(s, p, a)) continue;
                    ser.x.push_back(p);
                    ser.y.push_back(pick(a, 0.0));
                }
            }
            if (!ser.x.empty()) out.push_back(ser);
        }
        return out;
    };

    harness::svgPlot("benchmark/plots/time.svg",
                     "Q1 - Execution time vs processes",
                     "Number of MPI processes (P)", "t_algo (s, log scale)",
                     build([](const Agg& a, double) { return a.algo; }, false),
                     true, false);

    harness::svgPlot("benchmark/plots/speedup.svg",
                     "Q1 - Speed-up vs processes",
                     "Number of MPI processes (P)", "Speed-up  S(P)",
                     build([](const Agg&, double s) { return s; }, true),
                     false, true);

    {
        std::vector<Series> eff;
        for (auto& s : sizes) {
            Series ser;
            ser.name = s + " (" + shape[s] + ")";
            for (int p : procs) {
                auto it = speed.find(Key{s, p});
                if (it == speed.end()) continue;
                ser.x.push_back(p);
                ser.y.push_back(it->second / p);
            }
            if (!ser.x.empty()) eff.push_back(ser);
        }
        harness::svgPlot("benchmark/plots/efficiency.svg",
                         "Q1 - Efficiency vs processes",
                         "Number of MPI processes (P)", "Efficiency  E(P)",
                         eff, false, false);
    }

    harness::svgPlot(
        "benchmark/plots/comm_fraction.svg",
        "Q1 - Communication share of algorithm time",
        "Number of MPI processes (P)", "t_comm / t_algo",
        build([](const Agg& a, double) {
            return a.algo > 0 ? a.comm / a.algo : 0.0;
        }, false),
        false, false);

    std::printf("\n### Plots written\n\n");
    std::printf("- `benchmark/plots/time.svg`\n");
    std::printf("- `benchmark/plots/speedup.svg`\n");
    std::printf("- `benchmark/plots/efficiency.svg`\n");
    std::printf("- `benchmark/plots/comm_fraction.svg`\n");
    return 0;
}

File: HW2/Q1/benchmark/analyze.py
#!/usr/bin/env python3
"""
analyze.py -- turn benchmark/results.csv into the tables and plots for the Q1
report, in the layout of report_format.md.

Produces on stdout:
  * runtime table       (median t_algo, seconds)
  * speed-up table      S(P) = T1 / TP
  * efficiency table    E(P) = S(P) / P
  * communication breakdown: t_dist / t_compute / t_gather, and comm as % of
    t_algo -- this is the "communication vs computation" analysis the report
    format asks for

And writes benchmark/plots/{time,speedup,efficiency,comm_fraction}.png

The reported value per configuration is the MEDIAN over trials, which is
robust against a single slow run caused by another job on the node.

T1 is the MPI binary at P=1, not the sequential program, so the speed-up
measures parallel scaling of one implementation rather than comparing two
different programs.

Usage: python3 benchmark/analyze.py [results.csv]
"""

import csv
import os
import statistics
import sys
from collections import defaultdict

HERE = os.path.dirname(os.path.abspath(__file__))
CSV = sys.argv[1] if len(sys.argv) > 1 else os.path.join(HERE, "results.csv")
PLOTS = os.path.join(HERE, "plots")

METRICS = ("t_dist", "t_compute", "t_gather", "t_total")


def load(path):
    if not os.path.exists(path):
        sys.exit(f"no results file at {path}\nRun benchmark/run_bench.sh first.")
    rows = []
    with open(path) as f:
        for r in csv.DictReader(f):
            try:
                r["P"] = int(r["P"])
                for k in METRICS:
                    r[k] = float(r[k])
                r["t_comm"] = r["t_dist"] + r["t_gather"]
                r["t_algo"] = r["t_compute"] + r["t_comm"]
            except (ValueError, KeyError):
                continue
            rows.append(r)
    if not rows:
        sys.exit("results file contains no usable rows")
    return rows


def aggregate(rows):
    b = defaultdict(lambda: defaultdict(list))
    for r in rows:
        key = (r["size"], r["P"])
        for m in METRICS + ("t_comm", "t_algo"):
            b[key][m].append(r[m])
    return {k: {m: statistics.median(v) for m, v in met.items()}
            for k, met in b.items()}


def table(title, header, body, note=None):
    print(f"\n### {title}\n")
    if note:
        print(f"{note}\n")
    print("| " + " | ".join(header) + " |")
    print("|" + "|".join([":--"] + [":-:"] * (len(header) - 1)) + "|")
    for row in body:
        print("| " + " | ".join(row) + " |")


def main():
    rows = load(CSV)
    agg = aggregate(rows)

    sizes, shape = [], {}
    for r in rows:
        if r["size"] not in sizes:
            sizes.append(r["size"])
            shape[r["size"]] = f"{r['m']}x{r['n']}x{r['p']}"
    procs = sorted({r["P"] for r in rows})

    print("# Q1 Benchmark Analysis — Row-Row Matrix Multiplication")
    print(f"\nSource: `{os.path.relpath(CSV)}`  |  process counts: "
          f"{', '.join(map(str, procs))}  |  statistic: median over trials")
    print("\n`t_algo` = `t_compute` + `t_comm`, where "
          "`t_comm` = `t_dist` (Bcast dims + Scatterv A + Bcast B) + "
          "`t_gather` (Gatherv C).")

    table("Matrix sizes", ["Label", "m x n x p"],
          [[s, shape[s]] for s in sizes])

    # ---- runtime -----------------------------------------------------------
    body = []
    for s in sizes:
        cells = [s]
        for p in procs:
            a = agg.get((s, p))
            cells.append(f"{a['t_algo']:.4f}" if a else "--")
        body.append(cells)
    table("Runtime — median `t_algo` (seconds)",
          ["Input size"] + [f"P={p}" for p in procs], body)

    # ---- speed-up ----------------------------------------------------------
    speed = {}
    body = []
    for s in sizes:
        base = agg.get((s, 1))
        cells = [s]
        for p in procs:
            a = agg.get((s, p))
            if a and base and a["t_algo"] > 0:
                v = base["t_algo"] / a["t_algo"]
                speed[(s, p)] = v
                cells.append(f"{v:.2f}")
            else:
                cells.append("--")
        body.append(cells)
    table("Speed-up  S(P) = T1 / TP",
          ["Input size"] + [f"P={p}" for p in procs], body)

    # ---- efficiency --------------------------------------------------------
    body = []
    for s in sizes:
        cells = [s]
        for p in procs:
            v = speed.get((s, p))
            cells.append(f"{100.0 * v / p:.1f}%" if v else "--")
        body.append(cells)
    table("Efficiency  E(P) = S(P) / P",
          ["Input size"] + [f"P={p}" for p in procs], body)

    # ---- communication share ----------------------------------------------
    body = []
    for s in sizes:
        cells = [s]
        for p in procs:
            a = agg.get((s, p))
            if a and a["t_algo"] > 0:
                cells.append(f"{100.0 * a['t_comm'] / a['t_algo']:.1f}%")
            else:
                cells.append("--")
        body.append(cells)
    table("Communication share — `t_comm` as % of `t_algo`",
          ["Input size"] + [f"P={p}" for p in procs], body,
          note="Rising with P is expected: the Bcast of B moves n*p elements "
               "to every process regardless of P, while the computation per "
               "process falls as 1/P.")

    # ---- phase breakdown ---------------------------------------------------
    body = []
    for s in sizes:
        for p in procs:
            a = agg.get((s, p))
            if not a:
                continue
            body.append([s, str(p), f"{a['t_dist']:.4f}",
                         f"{a['t_compute']:.4f}", f"{a['t_gather']:.4f}",
                         f"{a['t_algo']:.4f}"])
    table("Phase breakdown (seconds, median)",
          ["Size", "P", "t_dist", "t_compute", "t_gather", "t_algo"], body)

    # ---- plots -------------------------------------------------------------
    try:
        import matplotlib
        matplotlib.use("Agg")
        import matplotlib.pyplot as plt
    except ImportError:
        print("\n(matplotlib not available - plots skipped)")
        return

    os.makedirs(PLOTS, exist_ok=True)

    def lineplot(fname, ylabel, title, valfn, ideal=False, logy=False):
        fig, ax = plt.subplots(figsize=(7, 4.5))
        for s in sizes:
            xs, ys = [], []
            for p in procs:
                v = valfn(s, p)
                if v is not None:
                    xs.append(p); ys.append(v)
            if xs:
                ax.plot(xs, ys, marker="o", label=f"{s} ({shape[s]})")
        if ideal:
            ax.plot(procs, procs, "k--", alpha=0.4, label="ideal (linear)")
        if logy:
            ax.set_yscale("log")
        ax.set_xlabel("Number of MPI processes (P)")
        ax.set_ylabel(ylabel)
        ax.set_title(title)
        ax.set_xticks(procs)
        ax.grid(alpha=0.3)
        ax.legend(fontsize=7)
        fig.tight_layout()
        path = os.path.join(PLOTS, fname)
        fig.savefig(path, dpi=150)
        plt.close(fig)
        return path

    made = [
        lineplot("time.png", "t_algo (s, log scale)",
                 "Q1 — Execution time vs processes",
                 lambda s, p: agg[(s, p)]["t_algo"] if (s, p) in agg else None,
                 logy=True),
        lineplot("speedup.png", "Speed-up  S(P)",
                 "Q1 — Speed-up vs processes",
                 lambda s, p: speed.get((s, p)), ideal=True),
        lineplot("efficiency.png", "Efficiency  E(P)",
                 "Q1 — Efficiency vs processes",
                 lambda s, p: (speed[(s, p)] / p) if (s, p) in speed else None),
        lineplot("comm_fraction.png", "t_comm / t_algo",
                 "Q1 — Communication share of algorithm time",
                 lambda s, p: (agg[(s, p)]["t_comm"] / agg[(s, p)]["t_algo"])
                 if (s, p) in agg and agg[(s, p)]["t_algo"] > 0 else None),
    ]

    print("\n### Plots written\n")
    for path in made:
        print(f"- `{os.path.relpath(path)}`")


if __name__ == "__main__":
    main()
